### PR TITLE
Documentation for data requirements of raw files

### LINKED ISSUE #106: [Feature Request]
As a developer

I need to know what columns are necessary for subsequent processing steps, especially structuring

so that I can convert and rename the right columns and determine if I am missing any requirements

Acceptance Criteria:

List of columns needed for structuring, their contents, units, data types, and standardized name

### PR DESCRIPTION
**Pull request by ardunn**
### Work in progress

This PR adds documentation to the readme to make it clear what column headers are required for the entire BEEP pipeline (primarily validation + structuring). The sources of truth of all these requirements are the validation yaml files  + the coversion yaml files; perhaps @patrickherring-TRI can advise on whether to use `*_conversion.yaml` or `schema-*.yaml` fields, as most validation `schema-*.yaml` seem to be subsets of the structuring `*_conversion.yaml` 

Closes #106

To do:
- [X] Arbin requirements (metadata + data)
- [X] Maccor requirements (data)
- [X] biologic requirements (data, some issues)
- [X] neware requirements (data, some issues)
- [X] indigo requirements (data, some issues)

Extra:
- [X] Added ToC to the readme, as it is getting long (maybe should consider dedicated documentation pages)
- [X] maccor_conversion.yaml was missing a data type for `dpt time` field; as per validation schema, should be `str`, so it was changed to that
- [X] fix imputation issue when various "nominally required" columns are missing, such as temperature and internal_resistance
- [X] fix `RawCyclerRun.from_neware_file` printing df data to stdout

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -241,8 +249,95 @@ $ run_model '{
...
+| `step_time` |   |   |   |   `float32` |
+| `step_index` |   |   |   |   `int16` |
+| `cycle_index` |   |   |   |   `int32` |
+| `current` |   |   |   |   `float32` |
+| `voltage` |   |   |   |   `float32` |
+| `charge_capacity` |   |   |   |   `float64` |
+| `discharge_capacity` |   |   |   |   `float64` |
+| `charge_energy` |   |   |   |   `float64` |
+| `discharge_energy` |   |   |   |   `float64` |
+| `dv/dt` |   |   |   |   `float32` |
```
Comment (patrickherring-TRI) [inline README.md:284]: I don't think that dv/dt is necessary for any processing down the line

File: README.md [hunk 2]
```diff
@@ -241,8 +249,95 @@ $ run_model '{
...
+| `internal_resistance` |   |   |   |   `float32` |
+| `temperature` |   |   |   |   `float32` |
```
Comment (patrickherring-TRI) [inline README.md:286]: I believe that temperature and internal resistance are optional.

**Comment by patrickherring-TRI** [comment]
@ardunn `*_conversion.yaml` should be the fields to pay attention to. We can reconcile with the `schema*.yaml` files when we have this portion done.

**Comment by ardunn** [comment]
> @ardunn `*_conversion.yaml` should be the fields to pay attention to. We can reconcile with the `schema*.yaml` files when we have this portion done.

Ok!

File: README.md (hunk 2 above)
Comment (ardunn) [reply README.md:286]: Hmm it seems Temperature is actually a required field for processedcycler run? Maybe not itentionally so. removing that column from the `2017-12-04_4_65C-69per_6C_CH29.csv` test file throws an error when ProcessedCyclerRun is instantiated via `pcr = ProcessedCyclerRun.from_raw_cycler_run(...)`

Comment (ardunn) [reply README.md:286]: A minimal script to reproduce:

```
from beep.structure import ProcessedCyclerRun, RawCyclerRun
import shutil
import pandas as pd


ARBIN_FILE = "/path/to/beep/beep/tests/test_files/2017-12-04_4_65C-69per_6C_CH29.csv"
ARBIN_METADATA = "/path/to/beep/beep/tests/test_files/2017-12-04_4_65C-69per_6C_CH29_Metadata.csv"


# create new input files without the temperature column
df = pd.read_csv(ARBIN_FILE)
metadf = pd.read_csv(ARBIN_METADATA)
df_removed = df.drop(columns=["Temperature"])
new_data_fname = ARBIN_FILE.replace(".csv", f"_no_Temperature.csv")
new_metadata_fname = ARBIN_METADATA.replace("_Metadata.csv", f"_no_Temperature_Metadata.csv")
df_removed.to_csv(new_data_fname, index=False)
shutil.copy(ARBIN_METADATA, new_metadata_fname)

# try to structure from new input files 
rcr = RawCyclerRun.from_arbin_file(new_data_fname)
pcr = ProcessedCyclerRun.from_raw_cycler_run(rcr)
```


Throws:

```
Traceback (most recent call last):
  File "/Users/ardunn/alex/tri/code/beep/alex_scripts/tmp.py", line 33, in <module>
    pcr = ProcessedCyclerRun.from_raw_cycler_run(rcr)
  File "/Users/ardunn/alex/tri/code/beep/beep/structure.py", line 1392, in from_raw_cycler_run
    diagnostic_available=diagnostic_available,
  File "/Users/ardunn/alex/tri/code/beep/beep/structure.py", line 289, in get_interpolated_cycles
    axis="voltage",
  File "/Users/ardunn/alex/tri/code/beep/beep/structure.py", line 234, in get_interpolated_steps
    resolution=resolution,
  File "/Users/ardunn/alex/tri/code/beep/beep/structure.py", line 1809, in get_interpolated_data
    df = dataframe.loc[:, columns]
  File "/Users/ardunn/alex/tri/code/beepenv/lib/python3.7/site-packages/pandas/core/indexing.py", line 1761, in __getitem__
    return self._getitem_tuple(key)
  File "/Users/ardunn/alex/tri/code/beepenv/lib/python3.7/site-packages/pandas/core/indexing.py", line 1288, in _getitem_tuple
    retval = getattr(retval, self.name)._getitem_axis(key, axis=i)
  File "/Users/ardunn/alex/tri/code/beepenv/lib/python3.7/site-packages/pandas/core/indexing.py", line 1953, in _getitem_axis
    return self._getitem_iterable(key, axis=axis)
  File "/Users/ardunn/alex/tri/code/beepenv/lib/python3.7/site-packages/pandas/core/indexing.py", line 1594, in _getitem_iterable
    keyarr, indexer = self._get_listlike_indexer(key, axis, raise_missing=False)
  File "/Users/ardunn/alex/tri/code/beepenv/lib/python3.7/site-packages/pandas/core/indexing.py", line 1552, in _get_listlike_indexer
    keyarr, indexer, o._get_axis_number(axis), raise_missing=raise_missing
  File "/Users/ardunn/alex/tri/code/beepenv/lib/python3.7/site-packages/pandas/core/indexing.py", line 1654, in _validate_read_indexer
    "Passing list-likes to .loc or [] with any missing labels "
KeyError: 'Passing list-likes to .loc or [] with any missing labels is no longer supported, see https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike'
```

If unable to reproduce, the problem is likely on my end

**Comment by ardunn** [comment]
@patrickherring-TRI the Arbin configuration is done. The fields marked "required" are those which fail in either `RawCyclerRun` or subsequent `ProcessedCyclerRun` when they are missing from the input file. The descriptions and units I couldn't find anywhere in the beep code, so I took them directly from the [Arbin manual](https://manualzz.com/doc/6665748/table-of-contents---arbin-instruments)

If this looks ok to go, I can proceed with the remaining cyclers (maccor, fastcharge, biologic, neware, indigo) in the same fashion

File: README.md (hunk 2 above)
Comment (ardunn) [reply README.md]: So it looks like we have two options:


#### Option 1: enforcing imputation across all cyclers in `RawCyclerRun.__init__`

We could remove that imputation from maccor specifically, and then enforce it with a kwarg to __init__ for all `RawCyclerRuns`. This might also solve the problem with Arbin (and yet unknown issues with the other cyclers). So like the `__init__` of `RawCyclerRun` could include an `inpute_missing` argument:

```
def __init__(self, data, metadata, eis=None, validate=False, filename=None, impute_missing=True):
```

Where, if True, `RawCyclerRun` will automatically nan the columns which are required for downstream processing but don't need actual values (e.g., Temperature). 

#### Option 2: Remove the need for temperature columns altogether

If temperature is not required for the processing steps themselves, we could just remove whatever common code needs it. Then we would not need to impute any of these "nominally required" columns which are not actually required for interpolation but cause errors when missing. From looking at that stacktrace it doesn't seem there is a direct way to fix it, but maybe someone who is more familiar with that code has thoughts?

Comment (patrickherring-TRI) [reply README.md]: I like Option 1 the best. I think that for most of the experiments, there should be temperature data and if there is not imputing with NaN makes the most sense.

Comment (ardunn) [reply README.md]: Ok, it turns out option 1 alone was not very robust so I wound up doing both.

Temperature was only nominally required (seemingly) because of a bad `.loc` which fails if it is missing, and separately because of `structure_dtypes.yaml` which implicitly enforced that field to be there. Ditto for internal_resistance. These issues are fixed with latest commit (though not sure if they will pass all tests yet)

**Comment by ardunn** [comment]
@patrickherring-TRI for Maccor, it seems all columns are nominally required because of the line

```
        data = data.astype(MACCOR_CONFIG["data_types"])
```

Whether or not they are functionally required downstream. There are likely similar cases for he other non-Arbin cyclers. 

I think piecewise fixing of these kinds of issues for all the `from_x_file methods` will draw this PR out substantially. I am just going to mark the columns required (whether they be nominal or truly, functionally required for processing) if the expected behavior breaks **for any reason** from a user perspective. We can change the underlying behavior down the line or in another PR.

File: README.md (hunk 2 above)
Comment (ardunn) [reply README.md]: Ok, I went with option 1. It is not a great fix, but will do in the interim.

**Review by patrickherring-TRI — APPROVED** [review]
LGTM, thanks for making those fixes!

**Comment by ardunn** [comment]
Some clarifications on this PR.


## Arbin
Seems all good to go. the fields have been systematically tested. Column descriptions are taken straight from the manufacturer manual.

## Maccor
Fields have been systematically tested. The internal workings of `RawCyclerRun.from_maccor_file` make any column missing nominally fail. This can be fixed, but not in this PR. I couldn't find a useful maccor manual, so I manually inferred the meanings of columns based on the names. This probably should be looked at quickly by someone to make sure the wording is right.

## Indigo
The fields for indigo have been systematically tested with a few caveats:
- The indigo conversion schema has no enforced datatypes, unlike Arbin for example.
- The indigo fields in the test files do not match the conversion schema file. ie. where is 'cell_coulomb_count_c' in the schema? I used the test file as the source of truth here since there is a Venn-diagra-like situation with the field names, and the fields in the conversion schema don't appear in any test files I saw.
Like maccor, I manually inferred the column meanings.


## Biologic
Biologic fields are have not been systematically tested for a few reasons:
- Biologic fails on `ProcessedCyclerRun`. I'm not sure if its supposed to, but there is no test to check against this. I'm guessing it has just not been implemented yet.
- The conversion schema is a subset of that in the original data file, so it's unclear what is actually "required". The schema also seems to be in a different format from other conversion schema (e.g., arbin).

- fastcharge and xtesladiag conversions seem to be extensions of arbin/maccor files, unclear how they work/what the requirements are

## Neware
Not tested systematically, for the reasons:

- The file format is just difficult to work with; i.e., what the column "headers" are is hard to parse out. The processing by `RawCyclerRun` estranges the actual "input dataframe" from the raw file enough that mapping precise requirements for the raw file would be hard for a developer. You seem to have to trust the Neware output is correct.
- Instead, I just assumed the conversion schema was 100% correct and users/developers will take the time to ensure the units and processing is correct either thru the Neware software or other means. So the source of truth here is just the conversion schema.
- That said, the conversion schema is missing data_types fields
- Additionally, some fields' keys in neware schema (`test_time`), are not in the raw test document (`neware_test.csv`). these were removed so only fields in both the schema and the test document are listed.

**Comment by ardunn** [comment]
@patrickherring-TRI tests pass but coveralls fails because coverage decreased by 0.04%
